timex/timex.go

```diff
@@ -0,0 +1,57 @@
+// Package timex should be called Rolex.
+package timex
+
+import (
+	"context"
+	"time"
+)
+
+// Ticker ...
+type Ticker struct {
+	C <-chan time.Time
+
+	tt   *time.Ticker
+	done chan bool
+	c    chan time.Time
+}
+
+// NewTicker ...
+func NewTicker(d time.Duration) *Ticker {
+	c := make(chan time.Time, 1)
+	ticker := Ticker{
+		C:    c,
+		tt:   time.NewTicker(d),
+		done: make(chan bool),
+		c:    c,
+	}
+
+	go func() {
+		for {
+			select {
+			case <-ticker.done:
+				return
+			case t := <-ticker.tt.C:
+				ticker.c <- t
+			}
+		}
+	}()
+
+	return &ticker
+}
+
+// Stop ...
+func (t *Ticker) Stop() {
+	t.tt.Stop()
+	t.done <- true
+	close(t.c)
+}
+
+// NewTickerWithContext ...
+func NewTickerWithContext(ctx context.Context, d time.Duration) *Ticker {
+	t := NewTicker(d)
+	go func() {
+		<-ctx.Done()
```

Comment:
It is possible that `context.Done()` returns `nil` if the context cannot be cancelled. In this case, this will leak a goroutine that will be blocked forever on a nil channel. I think you should only run the goroutine if `ctx.Done() != nil`.

Reply:
done()